geode-apis-compatible-with-redis/src/test/java/org/apache/geode/redis/internal/collections/OrderStatisticTreeTest.java

```diff
@@ -0,0 +1,580 @@
+      assertEquals(set.contains(i), tree.contains(i));
+    }
+  }
+
+  @Test
+  public void testIteratorBruteForce() {
+    for (int i = 0; i < 10_000; ++i) {
+      assertEquals(set.add(i), tree.add(i));
+    }
+
+    Iterator<Integer> iterator1 = set.iterator();
+    Iterator<Integer> iterator2 = tree.iterator();
+
+    long seed = System.nanoTime();
+    Random random = new Random(seed);
+
+    System.out.println("testIteratorBruteForce - seed: " + seed);
+
+    while (true) {
+      if (!iterator1.hasNext()) {
+        assertFalse(iterator2.hasNext());
+        break;
+      }
+
+      boolean toRemove = random.nextBoolean();
+
+      if (toRemove) {
+        try {
+          iterator1.remove();
+          iterator2.remove();
+        } catch (IllegalStateException ex) {
+          assertThatThrownBy(() -> iterator2.remove()).isInstanceOf(IllegalStateException.class);
+        }
+      } else {
+        assertEquals(iterator1.hasNext(), iterator2.hasNext());
+
+        if (iterator1.hasNext()) {
+          assertEquals(iterator1.next(), iterator2.next());
+        } else {
+          break;
+        }
+      }
+    }
+
+    assertEquals(set.size(), tree.size());
+    assertTrue(tree.isHealthy());
+    assertTrue(set.containsAll(tree));
+    assertTrue(tree.containsAll(set));
+  }
+
+  @Test
+  public void testIteratorConcurrentModification() {
+    for (int i = 0; i < 100; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> iterator1 = set.iterator();
+    Iterator<Integer> iterator2 = tree.iterator();
+
+    set.remove(10);
+    tree.remove(10);
+
+    assertEquals(iterator1.hasNext(), iterator2.hasNext());
+
+    try {
+      iterator1.next();
+      try {
+        iterator2.next();
+      } catch (ConcurrentModificationException ex) {
+        fail("iterator2 should NOT have thrown an exception.");
+      }
+    } catch (ConcurrentModificationException ex) {
+      assertThatThrownBy(() -> iterator2.next())
+          .isInstanceOf(ConcurrentModificationException.class);
+    }
+  }
+
+
+  @Test
+  public void testIteratorConcurrentRemove() {
+    for (int i = 10; i < 20; ++i) {
+      set.add(i);
+      tree.add(i);
+    }
+
+    Iterator<Integer> iterator1 = set.iterator();
+    Iterator<Integer> iterator2 = tree.iterator();
+
+    for (int i = 0; i < 4; ++i) {
+      iterator1.next();
+      iterator2.next();
+    }
+
+    // None of them contains 2, should not change the modification count.
+    set.remove(2);
+    tree.remove(2);
+
+    iterator1.remove();
+    iterator2.remove();
+
+    iterator1.next();
+    iterator2.next();
+
+    set.remove(12);
+    tree.remove(12);
+
+    // Both of them should throw.
+    assertThatThrownBy(() -> iterator1.remove())
+        .isInstanceOf(ConcurrentModificationException.class);
+    assertThatThrownBy(() -> iterator2.remove())
+        .isInstanceOf(ConcurrentModificationException.class);
+  }
+
+  @Test
+  public void testConcurrentOrIllegalStateOnRemove() {
```

Comment:
This test name seems to imply the test is doing something different from what it's actually doing. The test throws an `IllegalStateException` because `remove()` is called before `next()` has been called, not because of the call to `add()`. If `next()` is called on each iterator before the `add()` call, then the calls to `remove()` throw `ConcurrentModificationException`, which is what I think the test was originally intended to show.

Reply:
Yup, but the subsequent test does the `ConcurrentModificationException` test.